# Задача линейного программирования

In [20]:
from scipy.optimize import linprog
import time
start = time.time()
c = [-3,-4,-3,-1] 
b_ub = [80,480,130] 
A_ub = [[7,2,2,6], 
        [5,8,4,3],
        [2,4,1,8]] 
print (linprog(c, A_ub, b_ub))
stop = time.time()
print ("Время :")
print(stop - start)

     fun: -150.0
 message: 'Optimization terminated successfully.'
     nit: 2
   slack: array([  0., 200.,   0.])
  status: 0
 success: True
       x: array([ 0., 30., 10.,  0.])
Время :
0.0025069713592529297


Данная функция по умолчанию ищет минимум функции, соответственно нужно сделать коэффициенты целевой функции были отрицательными. Таким образом, максимум функции будет равен 150. x2 = 30, x3 = 10.

In [22]:
x1 = 40
x2 = 42
x3 = 37
x4 = x1 * 0.7 + x2 * 0.3
x5 = x1 * 0.6 + x3 * 0.4
x6 = x1 * 0.5 + x2 * 0.2 + x3 * 0.3

In [23]:
print (x1, x2, x3, x4, x5, x6)

40 42 37 40.6 38.8 39.5


# Решение транспортной задачи

In [6]:
from scipy.optimize import linprog
import time
start = time.time()
c = [5,8,7,10,3,4,2,2,5,6,7,3,5,9,2]
b_ub = [200,450,250] 
A_ub = [[1,1,1,1,1,0,0,0,0,0,0,0,0,0,0],
               [0,0,0,0,0,1,1,1,1,1,0,0,0,0,0],
               [0,0,0,0,0,0,0,0,0,0,1,1,1,1,1]] 
b_eq = [100,125,325,250,100] 
A_eq = [[1,0,0,0,0,1,0,0,0,0,1,0,0,0,0],
               [0,1,0,0,0,0,1,0,0,0,0,1,0,0,0],
               [0,0,1,0,0,0,0,1,0,0,0,0,1,0,0],
               [0,0,0,1,0,0,0,0,1,0,0,0,0,1,0],
               [0,0,0,0,1,0,0,0,0,1,0,0,0,0,1]] 
print(linprog(c, A_ub, b_ub, A_eq, b_eq))
stop = time.time()
print ("Время :")
print(stop - start)

     fun: 3450.0
 message: 'Optimization terminated successfully.'
     nit: 15
   slack: array([0., 0., 0.])
  status: 0
 success: True
       x: array([100.,   0.,   0.,   0., 100.,   0.,   0., 200., 250.,   0.,   0.,
       125., 125.,   0.,   0.])
Время :
0.010025978088378906


In [7]:
from scipy.optimize import linprog
import time
start = time.time()
c = [3,4,9,18,9,6,
    16,8,12,13,20,4,
    8,6,13,1,6,9,
    16,9,6,8,1,11,
    8,12,17,5,3,5,
    2,9,1,10,5,17]
b_ub = [200,450,250] 
A_ub = [[1,1,1,1,1,0,0,0,0,0,0,0,0,0,0],
               [0,0,0,0,0,1,1,1,1,1,0,0,0,0,0],
               [0,0,0,0,0,0,0,0,0,0,1,1,1,1,1]] 
b_eq = [100,125,325,250,100] 
A_eq = [[1,0,0,0,0,1,0,0,0,0,1,0,0,0,0],
               [0,1,0,0,0,0,1,0,0,0,0,1,0,0,0],
               [0,0,1,0,0,0,0,1,0,0,0,0,1,0,0],
               [0,0,0,1,0,0,0,0,1,0,0,0,0,1,0],
               [0,0,0,0,1,0,0,0,0,1,0,0,0,0,1]] 
print(linprog(c, A_ub, b_ub, A_eq, b_eq))
stop = time.time()
print ("Время :")
print(stop - start)

     fun: 3450.0
 message: 'Optimization terminated successfully.'
     nit: 15
   slack: array([0., 0., 0.])
  status: 0
 success: True
       x: array([100.,   0.,   0.,   0., 100.,   0.,   0., 200., 250.,   0.,   0.,
       125., 125.,   0.,   0.])
Время :
0.009024381637573242


In [8]:
import pip
pip.main(['install', 'scipy'])

AttributeError: module 'pip' has no attribute 'main'

In [9]:
pip.main(['install', '--user', 'scipy'])

AttributeError: module 'pip' has no attribute 'main'

In [25]:
c=[1,2,3,4,5,6,7]
x=[9,8,7,6,5,4,3]
k=0

for i in range(len(c)):
    k += c[i]*x[i]
print(k)

140
